test/test_infer.rb

```diff
@@ -8,6 +8,60 @@
 require 'rdl'
 require 'types/core'
 
+class CompareTypeVariables
+  def initialize
+    @var_lookup = {}
+  end
+
+  def compare_methods(expected, given)
+    expected.args.length == given.args.length &&
+      expected.args.zip(given.args).all? { |a| compare(a[0], a[1]) } &&
+      compare(expected.block, given.block) &&
+      compare(expected.ret, given.ret)
+  end
+
+  def compare_structural(expected, given)
+    expected.methods.keys == given.methods.keys &&
+      expected.methods.values.zip(given.methods.values).all? { |a| compare(a[0], a[1]) }
+  end
+
+  def compare_hash(expected, given)
+    expected.elts.keys == given.elts.keys &&
+      expected.elts.values.zip(given.elts.values).all? { |a| compare(a[0], a[1]) }
+  end
+
+  def compare_vartype(expected, given)
+    name_sym = expected.name.to_sym
+    return @var_lookup[name_sym] == given if @var_lookup.key? name_sym
+
+    @var_lookup[name_sym] = given
+    true
+  end
+
+  def compare(expected, given)
+    if expected.class == given.class
+
+      case expected
+      when RDL::Type::MethodType
+        return compare_methods(expected, given)
+
+      when RDL::Type::StructuralType
+        return compare_structural(expected, given)
+
+      when RDL::Type::FiniteHashType
+        return compare_hash(expected, given)
+
+      when RDL::Type::VarType
+        return compare_vartype(expected, given)
+      end
+
+      return expected == given
```

Comment:
So this doesn't cover all of the types that might have type variables somewhere in their hierarchy, but this is enough to pass the tests. I'm not sure if we want this to be exhaustive at the start, or if we want to only grow it as we add more test cases.